cpp/src/arrow/compute/kernels/select_k_test.cc

```diff
@@ -0,0 +1,809 @@
+  this->AssertSelectKJson(R"(["testing", null, "nth", "for", null, "strings"])", 6);
+}
+
+template <typename ArrowType>
+class TestSelectKRandom : public TestSelectKBase<ArrowType> {
+ public:
+  std::shared_ptr<DataType> type_singleton() override {
+    EXPECT_TRUE(0) << "shouldn't be used";
+    return nullptr;
+  }
+};
+
+using SelectKableNumericAndTemporal =
+    ::testing::Types<UInt8Type, UInt16Type, UInt32Type, UInt64Type, Int8Type, Int16Type,
+                     Int32Type, Int64Type, FloatType, DoubleType, Date32Type, Date64Type,
+                     TimestampType, Time32Type, Time64Type>;
+
+using SelectKableTypes =
+    ::testing::Types<UInt8Type, UInt16Type, UInt32Type, UInt64Type, Int8Type, Int16Type,
+                     Int32Type, Int64Type, FloatType, DoubleType, Decimal128Type,
+                     StringType>;
+
+TYPED_TEST_SUITE(TestSelectKRandom, SelectKableTypes);
+
+TYPED_TEST(TestSelectKRandom, RandomValues) {
+  Random<TypeParam> rand(0x61549225);
+  int length = 100;
+  for (auto null_probability : {0.0, 0.1, 0.5, 1.0}) {
+    // Try n from 0 to out of bound
+    for (int n = 0; n <= length; ++n) {
+      auto array = rand.Generate(length, null_probability);
+      this->AssertTopKArray(array, n);
+      this->AssertBottomKArray(array, n);
+    }
+  }
+}
+
+template <SortOrder order>
+struct TestSelectKWithChunkedArray : public ::testing::Test {
+  TestSelectKWithChunkedArray()
+      : sizes_({0, 1, 2, 4, 16, 31, 1234}),
+        null_probabilities_({0.0, 0.1, 0.5, 0.9, 1.0}) {}
+
+  void Check(const std::shared_ptr<DataType>& type,
+             const std::vector<std::string>& values, int64_t k,
+             const std::string& expected) {
+    std::shared_ptr<Array> actual;
+    ASSERT_OK(this->DoSelectK(type, values, k, &actual));
+    ASSERT_ARRAYS_EQUAL(*ArrayFromJSON(type, expected), *actual);
+  }
+
+  void Check(const std::shared_ptr<DataType>& type,
+             const std::shared_ptr<ChunkedArray>& values, int64_t k,
+             const std::string& expected) {
+    ASSERT_OK_AND_ASSIGN(auto actual, SelectK<order>(*values, k));
+    ValidateOutput(actual);
+    ASSERT_ARRAYS_EQUAL(*ArrayFromJSON(type, expected), *actual);
+  }
+
+  Status DoSelectK(const std::shared_ptr<DataType>& type,
+                   const std::vector<std::string>& values, int64_t k,
+                   std::shared_ptr<Array>* out) {
+    ARROW_ASSIGN_OR_RAISE(*out, SelectK<order>(*(ChunkedArrayFromJSON(type, values)), k));
+    ValidateOutput(*out);
+    return Status::OK();
+  }
+  std::vector<int32_t> sizes_;
+  std::vector<double> null_probabilities_;
+};
+
+template <typename ArrowType>
+struct TestTopKWithChunkedArray
+    : public TestSelectKWithChunkedArray<SortOrder::Descending> {
+  std::shared_ptr<DataType> type_singleton() {
+    return default_type_instance<ArrowType>();
+  }
+};
+
+TYPED_TEST_SUITE(TestTopKWithChunkedArray, SelectKableNumericAndTemporal);
```

Comment:
For chunked array tests, you could get it "for free" by slicing a non-chunked array and building a chunked array to test. This is what the scalar kernel tests do. That way, you don't have to duplicate test cases.

Reply:
You could imagine doing the same thing for tables/record batches.